File: app/src/main/java/com/gooseco/technicalrulesquiz/data/repository/QuizRepository.kt
package com.gooseco.technicalrulesquiz.data.repository

import android.content.Context
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import com.gooseco.technicalrulesquiz.data.model.QuizQuestion
import java.io.IOException

class QuizRepository(private val context: Context) {

    private var cachedQuestions: List<QuizQuestion>? = null

    /**
     * Load all questions from the JSON asset file
     */
    private fun loadQuestionsFromJson(): List<QuizQuestion> {
        cachedQuestions?.let { return it }

        return try {
            val jsonString = context.assets.open("questions.json").bufferedReader().use { it.readText() }
            val gson = Gson()
            val questionData = gson.fromJson(jsonString, QuestionData::class.java)
            val questions = questionData?.questions ?: emptyList()
            cachedQuestions = questions
            questions
        } catch (e: Exception) {
            e.printStackTrace()
            emptyList()
        }
    }

    /**
     * Get all validated questions (status == "validated")
     * These are the questions that appear in the app quiz
     */
    fun getAllQuestions(): List<QuizQuestion> {
        return loadQuestionsFromJson().filter { it.status == "validated" }
    }

    /**
     * Get random validated questions for a quiz session
     */
    fun getRandomQuestions(count: Int): List<QuizQuestion> {
        val validatedQuestions = getAllQuestions()
        return validatedQuestions.shuffled().take(count)
    }

    /**
     * Get validated questions filtered by section
     */
    fun getQuestionsBySection(section: String): List<QuizQuestion> {
        return getAllQuestions().filter { it.ruleReference.section == section }
    }

    /**
     * Get all questions regardless of status (for debugging/admin purposes)
     */
    fun getAllQuestionsUnfiltered(): List<QuizQuestion> {
        return loadQuestionsFromJson()
    }

    /**
     * Get questions by status
     */
    fun getQuestionsByStatus(status: String): List<QuizQuestion> {
        return loadQuestionsFromJson().filter { it.status == status }
    }

    /**
     * Get count of questions by status
     */
    fun getQuestionCountByStatus(): Map<String, Int> {
        val questions = loadQuestionsFromJson()
        return questions.groupBy { it.status }.mapValues { it.value.size }
    }

    /**
     * Clear the cache to reload questions from JSON
     */
    fun clearCache() {
        cachedQuestions = null
    }

    /**
     * Data class to match the JSON structure
     */
    private data class QuestionData(
        @SerializedName("version") val version: String,
        @SerializedName("lastUpdated") val lastUpdated: String,
        @SerializedName("questions") val questions: List<QuizQuestion>
    )
}

File: ios/PowerliftingQuiz/Sources/Data/QuizRepository.swift
import Foundation

@Observable
final class QuizRepository {
    private(set) var allValidatedQuestions: [QuizQuestion] = []
    private var loaded = false

    func loadIfNeeded() {
        guard !loaded else { return }
        loaded = true

        guard let url = Bundle.main.url(forResource: "questions", withExtension: "json"),
              let data = try? Data(contentsOf: url) else {
            return
        }

        let decoded = try? JSONDecoder().decode(QuestionData.self, from: data)
        allValidatedQuestions = (decoded?.questions ?? []).filter { $0.status == "validated" }
    }

    func randomQuestions(count: Int) -> [QuizQuestion] {
        Array(allValidatedQuestions.shuffled().prefix(count))
    }

    var availableSections: [String] {
        Array(Set(allValidatedQuestions.map(\.ruleReference.section))).sorted()
    }

    var totalValidated: Int { allValidatedQuestions.count }
}

// MARK: - Private

private struct QuestionData: Codable {
    let version: String
    let lastUpdated: String
    let questions: [QuizQuestion]
}
